plane/src/types/backend_state.rs

```diff
@@ -0,0 +1,150 @@
+
+#[derive(Clone, Copy, Serialize, Deserialize, Debug, PartialEq, PartialOrd)]
+pub enum BackendStatus {
+    /// The backend has been scheduled to a drone, but has not yet been acknowledged.
+    /// This status is only assigned by the controller; the drone will never assign it by definition.
+    Scheduled,
+
+    /// The backend has been assigned to a drone, which is now responsible for loading its image.
+    Loading,
+
+    /// Telling Docker to start the container.
+    Starting,
+
+    /// Wait for the backend to be ready to accept connections.
+    Waiting,
+
+    /// The backend is listening for connections.
+    Ready,
+
+    /// The backend has been sent a SIGTERM, either because we sent it or the user did,
+    /// and we are waiting for it to exit.
+    /// Proxies should stop sending traffic to it, but we should not yet release the key.
+    Terminating,
+
+    /// The backend has exited or been swept.
+    Terminated,
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize, Debug, PartialEq)]
+pub enum TerminationKind {
+    Soft,
+    Hard,
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct BackendState {
+    pub status: BackendStatus,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub address: Option<SocketAddr>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub termination: Option<TerminationKind>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub reason: Option<TerminationReason>,
+
+    /// The last status before the backend entered Terminating or Terminated.
+    pub last_status: Option<BackendStatus>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub exit_code: Option<i32>,
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize, Debug, PartialEq)]
+pub enum TerminationReason {
+    Swept,
+    External,
+    KeyExpired,
+}
+
+impl BackendState {
+    pub fn to_loading(&self) -> BackendState {
+        BackendState {
+            status: BackendStatus::Loading,
+            ..self.clone()
+        }
+    }
+
+    pub fn to_starting(&self) -> BackendState {
+        BackendState {
+            status: BackendStatus::Starting,
+            ..self.clone()
+        }
+    }
+
+    pub fn to_waiting(&self, address: SocketAddr) -> BackendState {
+        BackendState {
+            status: BackendStatus::Waiting,
+            address: Some(address),
+            ..self.clone()
+        }
+    }
+
+    pub fn to_ready(&self) -> BackendState {
+        BackendState {
+            status: BackendStatus::Ready,
+            ..self.clone()
+        }
+    }
+
+    pub fn to_terminating(
+        &self,
+        termination: TerminationKind,
+        reason: TerminationReason,
+    ) -> BackendState {
+        BackendState {
+            status: BackendStatus::Terminating,
+            termination: Some(termination),
+            reason: Some(reason),
+            last_status: Some(self.status),
+            address: None,
+            exit_code: None,
```

Comment:
One tricky part here is that there are really two types of termination metadata: metadata about why a termination signal was sent (`termination` and `reason`), and metadata about the termination itself (`exit_code`). `last_status` can be populated either when transitioning into `Terminating` or, if `Terminating` is skipped, when entering `Terminated`.

I do like the idea of reducing the state space. Maybe this should be an `enum` where each possible `BackendStatus` has a variant with only the valid fields for that status? We could have a `.status()` function that would return the associated `BackendStatus` enum, so we'd still have access to it as a status for things that need that.

Reply:
> Maybe this should be an `enum` where each possible `BackendStatus` has a variant with only the valid fields for that status?

I like this! I think this would make the BackendStatus struct a lot more understandable and clarify to the reader what the state space actually encompasses.